client.go

```diff
@@ -74,6 +78,244 @@ func SelectOneOf(protos []string, rwc io.ReadWriteCloser) (string, error) {
 	return "", ErrNotSupported
 }
 
+// Performs protocol negotiation with the simultaneous open extension; the returned boolean
+// indicator will be true if we should act as a server.
+func SelectWithSimopen(protos []string, rwc io.ReadWriteCloser) (string, bool, error) {
+	if len(protos) == 0 {
+		return "", false, ErrNoProtocols
+	}
+
+	werrCh := make(chan error, 1)
+	go func() {
+		var buf bytes.Buffer
+		delimWrite(&buf, []byte(ProtocolID))
+		delimWrite(&buf, []byte("iamclient"))
+		delimWrite(&buf, []byte(protos[0]))
+		_, err := io.Copy(rwc, &buf)
+		werrCh <- err
+	}()
+
+	err := readMultistreamHeader(rwc)
+	if err != nil {
+		return "", false, err
+	}
+
+	tok, err := ReadNextToken(rwc)
+	if err != nil {
+		return "", false, err
+	}
+
+	if err = <-werrCh; err != nil {
+		return "", false, err
+	}
+
+	switch tok {
+	case "iamclient":
+		// simultaneous open
+		return simOpen(protos, rwc)
+
+	case "na":
+		// client open
+		proto, err := clientOpen(protos, rwc)
+		if err != nil {
+			return "", false, err
+		}
+
+		return proto, false, nil
+
+	default:
+		return "", false, errors.New("unexpected response: " + tok)
+	}
+}
+
+func clientOpen(protos []string, rwc io.ReadWriteCloser) (string, error) {
+	// check to see if we selected the pipelined protocol
+	tok, err := ReadNextToken(rwc)
+	if err != nil {
+		return "", err
+	}
+
+	switch tok {
+	case protos[0]:
+		return tok, nil
+	case "na":
+		// try the other protos
+		for _, p := range protos[1:] {
+			err = trySelect(p, rwc)
+			switch err {
+			case nil:
+				return p, nil
+			case ErrNotSupported:
+			default:
+				return "", err
+			}
+		}
+
+		return "", ErrNotSupported
+	default:
+		return "", errors.New("unexpected response: " + tok)
+	}
+}
+
+func simOpen(protos []string, rwc io.ReadWriteCloser) (string, bool, error) {
+	retries := 3
+
+again:
+	mynonce := make([]byte, 32)
+	_, err := rand.Read(mynonce)
+	if err != nil {
+		return "", false, err
+	}
+
+	werrCh := make(chan error, 1)
+	go func() {
+		myselect := []byte("select:" + base64.StdEncoding.EncodeToString(mynonce))
+		err := delimWriteBuffered(rwc, myselect)
+		werrCh <- err
+	}()
+
+	var peerselect string
+	for {
+		tok, err := ReadNextToken(rwc)
+		if err != nil {
+			return "", false, err
+		}
+
+		// this skips pipelined protocol negoatiation
+		// keep reading until the token starts with select:
+		if strings.HasPrefix(tok, "select:") {
```

Comment:
@Stebalien I am not sure what you mean here.

Are you talking about the exact scenario where a peer sends "mutlsitream"|"iamclient"|"select:"|"select:nonce"

where that first "select:" is a security protocol the user wants to negotiate ? I agree this scenario will create problems. 

But, is this the ONLY scenario you have in mind ? Does this really warrant fixing ?

Reply:
Yes, that's the scenario I have in mind. My concern isn't really "fixing a bug" but rather "simplifying the spec". If we go with the current approach, the multistream-select spec would need to say "protocols beginning with select: are reserved" and "iamclient" would become an integral part of the spec. If we simply skip _exactly_ one protocol, we can define "iamclient" to be it's _own_ protocol.

The "iamclient" (maybe `/bootstrap/simultaneous`?) protocol:

1. Send exactly one preferred protocol.
2. Send exactly one `select:nonce` message to pick the winner.
3. The winner recursively uses multistream-select to pick the final protocol. NOTE: this would require sending a new multistream header as well.

The beauty of this approach is that there's nothing inherently special about the "iamclient" protocol from a _spec_ standpoint. The _implementation_ can take advantage of the multistream-select ambiguity and pipeline "iamclient" with the "preferred protocol" from step 1, but that's not required. A simpler implementation could just:

1. Negotiate one of "iamclient" or a set of other security transports (one round-trip per step).
2. If "iamclient" is selected, send the preferred protocol and a nonce.
3. Once the winner is picked, the winner negotiates their chosen protocol, potentially using the "preferred" protocol.

This version takes an extra round-trip but can be implemented as two compossible protocols. Later, as the libp2p implementation matures, it can be re-implemented to remove the extra round-trip without changing the spec.

Reply:
TL;DR: it allows us to write the spec as follows:

1. Multistream protocol.
2. "iamclient" protocol for selecting a initiator when a simultaneous connect happens.
3. Optional: oh, by the way, you can combine these as follows ... to save a round-trip.

Reply:
@Stebalien This is done.